### ISSUE TITLE
Consider unified influxdb_iox_client Client structure

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#3007]
# Use case
This ticket is sparked by a discussion with @pauldix, @crepererum  and others, related to some work on https://github.com/influxdata/influxdb_iox/issues/2980 to make the `influxdb_iox_client` easy to work with. 

# Proposal:
The proposal would be to have a single `Client` structure from which various api calls would be made

# Current behavior
## Rust `influxdb_iox_client` `Client`s 
Each gRPC service maps to a single Rust `Client` client side structure type. To see how this looks, here is an example from the end to end tests. To close a partition via the management API and then wait for it to be complete, you use two different clients objects (that use the same underlying gRPC channel)

```rust
  use influxdb_iox_client::*;

  let connection = Builder::default()
    .build("http://127.0.0.1:8082")
    .await
    .expect("client should be valid");

  // start the close partition operation with a "management" client
  let management_client = management::Client::new(connection.clone())
  let iox_operation = management_client
        .close_partition_chunk(&db_name, table_name, partition_key, chunk_id)
        .await
        .unwrap()

    // wait for the job to be done using an "operations" client
    let operations_client = operations::Client::new(connection.clone());
    operations_client()
        .wait_operation(
            iox_operation.operation.id(),
            Some(std::time::Duration::from_secs(1)),
        )
        .await
        .unwrap();
```

## Golang `influxdb` client

The split into different client side structures is fairly similar to how the golang client is structured: https://pkg.go.dev/github.com/influxdata/influxdb-client-go/v2#Client.WriteAPI, though of course the terms are different.

* IOx Rust client `Connection` == Golang `Client`
* IOx Rust client `Client` == Golang `API` (e.g. WriteAPI())

## Rust InfluxDB client

(note This library seems very incomplete)

The the structure in https://docs.rs/influxdb/0.4.0/influxdb/struct.Client.html appears to imply it will have a single client struct, like the alternative option presented below

# Proposed Behavior
A common pattern in some other languages is a unified client -- so the previous example might look like:

```rust
  use influxdb_iox_client::*;

  let connection = Builder::default()
    .build("http://127.0.0.1:8082")
    .await
    .expect("client should be valid");

  // start the close partition and wait for the job using the same 'client' object
  let client = Client::new(connection.clone())
  let iox_operation = client
        .close_partition_chunk(&db_name, table_name, partition_key, chunk_id)
        .await
        .unwrap()

    client.wait_operation(
            iox_operation.operation.id(),
            Some(std::time::Duration::from_secs(1)),
        )
        .await
        .unwrap();
```

# Alternatives Considered:
Perhaps perhaps rename the structures to better align with the Golang client:
`Channel` --> `Client` 
`Client` --> `API`

**Comment by crepererum** [comment]
I like the `Channel->Client` and `Client->API` renaming. On top of that, I would propose the following:

While we have multiple APIs, users will usually interact w/ a dedicated server type (see #2962 ). Users might find it difficult to do the "which type implements which API" in their head. So we could use a trait per API (that basically does the full implementation) and use `impl`s to map them to server types. Something like this:

```rust
// base trait for all clients
trait Client {
    fn grpc_channel(&self) -> ...;
}

// generic client for untyped access and that is constructed by `Builder`
struct GenericClient {...}
impl Client for GenericClient {...}

// different APIs
trait DatabaseAPI: Client {
    // full implementation goes here
}

trait DeploymentAPI: Client {
    // full implementation goes here
}

trait RouterAPI: Client {
    // full implementation goes here
}

trait WriteAPI: Client {
    // full implementation goes here
}

// DB server
struct DatabaseServer {...}
impl From<GenericClient> for DatabaseServer {...}
impl Client for DatabaseServer {...}

// router server
struct RouterServer {...}
impl From<GenericClient> for RouterServer {...}
impl Client For RouterServer {...}

// now we API traits only act as markers
impl DatabaseAPI for GenericClient {}
impl DeploymentAPI for GenericClient {}
impl RouterAPI for GenericClient {}
impl WriteAPI for GenericClient {}

impl DatabaseAPI for DatabaseServer {}
impl DeploymentAPI for DatabaseServer {}
impl WriteAPI for DatabaseServer {}

impl DeploymentAPI for RouterServer {}
impl RouterAPI for RouterServer {}
impl WriteAPI for RouterServer {}
```

**Comment by tustvold** [comment]
Definitely agree with the renaming, consistency is good imo :smile: 

As for the trait plumbing, I like it in theory but wonder if it might be a bit premature to invest significant amounts of time into a client implementation. If I understand this suggestion correctly, it would result in a lot of additional plumbing for every method addition/alteration to the APIs which seems somewhat unfortunate? The amount of plumbing for manipulating the APIs is already a bit tiresome, I'm not particularly excited about introducing more at this stage in development where change is likely?

I don't know I'm very much in the camp of, auto-generated code may not be perfect, but it's also code you don't have to maintain...

**Comment by alamb** [comment]
@crepererum I like the trait idea -- and I also agree that any major investment in client code at the moment is likely premature. Let's see where @pauldix 's input comes in and then decide what to do

**Comment by crepererum** [comment]
> If I understand this suggestion correctly, it would result in a lot of additional plumbing for every method addition/alteration to the APIs which seems somewhat unfortunate?

No, because the methods are implemented on the API traits once and can then be reused for every server type. That's not more per-method pluming then we currently need. I've updated the example to make that a bit more clear.
